### ISSUE TITLE
Use defaults for undefined arguments in shorthand call to swal

### ISSUE THREAD

**Issue by zenflow** [sweetalert2/sweetalert2#991]
I found some unexpected behavior when working on #979...

[jsfiddle](https://jsfiddle.net/ad3quksn/745/?utm_source=website&utm_medium=embed&utm_campaign=ad3quksn)

```js
main()
async function main () {
  swal.setDefaults({type: 'question'})
  await swal({ title: 'I am the question type...'})
  await swal('but I am not?')
}
```

I would expect the second swal to also have `type` of `'question'`, but it does not.

It's due to this line of code, which assigns the values to `params` even when they are `undefined`..

https://github.com/sweetalert2/sweetalert2/blob/13d90857418f69094c7c259596323cd510b6c970/src/sweetalert2.js#L104

I propose that we treat `undefined` arguments as such, and use the parameter defaults in these cases. What do y'all think?

I'm not sure if this would be considered a fix or a breaking change. I mean, I'm pretty sure it's technically a breaking change, but bug fixes usually are. How likely is it that people are relying on the current behavior?

**Comment by limonte** [comment]
> I propose that we treat undefined arguments as such, and use the parameter defaults in these cases. What do y'all think?

Agree.

> I'm not sure if this would be considered a fix or a breaking change. I mean, I'm pretty sure it's technically a breaking change, but bug fixes usually are. How likely is it that people are relying on the current behavior?

Bug-fix IMO.

**Comment by zenflow** [comment]
> Bug-fix IMO.

Good. Yeah I agree.

I was thinking of something like this:

```js
function argsToParams(args) {
  const params = {}
  ;['title', 'html', 'type'].forEach((name, index) => {
    if (args[index] !== undefined) {
      params[name] = args[index]
    }
  })
  return params
}
```

I'll put in another PR after #979 settles

**Comment by toverux** [comment]
Yep, bug fix.

I'm not a micro-optimization guru nor I'm obsessed with performance but this part looks a bit suboptimal to me: 
> ['title', 'html', 'type'].forEach((name, index) => {

Instantiation of an array + 4 userland calls on the stack (plus the internal mechanics of forEach).

If you like this pattern like me (I won't blame you if you don't):

```js
args.title !== undefined && (params.title = args.title);
args.html !== undefined && (params.html = args.html);
args.type !== undefined && (params.type = args.type);
```

Not elegant but readable, the intent is clearer, and it's faster by a few microseconds :smile:

**Comment by zenflow** [comment]
@toverux I think micro-optimization on this library at this point doesn't make any sense, since, well, we have a lot of macro-optimization to make, e.g. not storing state almost entirely in the dom accessing the dom (slow) all the time like the [`elementByClass` I just fixed here](https://github.com/sweetalert2/sweetalert2/pull/990/commits/22057971f6195de53bb23beecd6156a73c727e00). I fixed a good chunk of it in PR #969, but lots remains.

Not to mention that our lib isn't really CPU intensive (our code isn't running very much of the time), JS engines are really fast (function calls are not that expensive), and [I think] elegant code is generally important (it's well-worth the cost).

I do have a preference for readability and maintainability over performance, but I think for this particular project it makes objectively more sense.

TBH I'd rather open up a new PR addressing the dom accessing issue than start making our code WET in the name of micro-optimizations.

**Comment by limonte** [comment]
I like @toverux's code not because of micro-optimization, but because of readability. It's dead simple, even a zombie will understand it :D

> Not elegant but readable

<3

PS. not related to this case, but I kile this joke by Wes Bos: https://youtu.be/9YkUCxvaLEk?t=10m4s

**Comment by toverux** [comment]
@zenflow You are very, _very_, right, especially concerning DOM operations. I was just saying that this [].forEach is just uselessly complex and "expensive", without bringing much readability (to me, of course). It just looks like "let's use a forEach and use lambdas, because lambdas" :)

**Comment by zenflow** [comment]
> It just looks like "let's use a forEach and use lambdas, because lambdas" :)

It's more like let's use a `forEach` because here's something I want to do once for each of these. I don't want to repeat myself. Btw, `Array.prototype.forEach` is not functional programming because it has no return value and we use it with functions that have side effects.

I don't see how it is complex. "Do this for each of these." It's simpler and more concise than your code example. More expensive, yes, barely, and worth the expense. 

It would be cool to use `for .. of` but we can't because of this #942.

I don't feel comfortable copy-pasting code for the sake of reducing function calls.

**Comment by toverux** [comment]
Aha, it's amusing to see that we always agree on complex subjects and general direction but always fail to find a compromise on the smallest details :laughing: 

> I don't see how it is complex.

> I don't feel comfortable copy-pasting code for the sake of reducing function calls.

It's not code duplication at this point :) You are a bit extreme here. It's just a small, pure-JS pattern applied to three variables. Our brain likes those visual layouts and patterns. Those LoCs don't need to be fully read, parsed and analyzed.
At the contrary, the array, the function, the loop, the array access syntax, they make this code _look_ complicated. I need to _read_ those lines when scanning the code.

But don't worry, I'm not asking you to change it if you don't agree, I don't care that much about that. I'm just saying that I would have not written it like that. And I'm sorry to have started another unproductive debate :smile:

**Comment by zenflow** [comment]
Released in version 7.14.0
